src/ray/core_worker/reference_count.cc

```diff
@@ -323,37 +330,43 @@ void ReferenceCounter::UpdateFinishedTaskReferences(
   RemoveSubmittedTaskReferences(argument_ids, release_lineage, deleted);
 }
 
-void ReferenceCounter::ReleaseLineageReferences(
-    const std::vector<ObjectID> &argument_ids) {
-  absl::MutexLock lock(&mutex_);
-  ReleaseLineageReferencesInternal(argument_ids);
-}
+int64_t ReferenceCounter::ReleaseLineageReferences(ReferenceTable::iterator ref) {
+  int64_t lineage_bytes_evicted = 0;
+  std::vector<ObjectID> argument_ids;
+  if (on_lineage_released_ && ref->second.owned_by_us) {
+    RAY_LOG(DEBUG) << "Releasing lineage for object " << ref->first;
+    lineage_bytes_evicted += on_lineage_released_(ref->first, &argument_ids);
+    // The object is still in scope by the application and it was
+    // reconstructable with lineage. Mark that its lineage has been evicted so
+    // we can return the right error during reconstruction.
+    if (!ref->second.OutOfScope(lineage_pinning_enabled_) &&
+        ref->second.is_reconstructable) {
+      ref->second.lineage_evicted = true;
+      ref->second.is_reconstructable = false;
```

Comment:
Are there cases where we cannot reconstruct the object although lineage is not evicted?

Reply:
I wonder if we can consolidate these two flags. If they can't can you add comments when it is not reconstructable in the `is_reconstructable ` docstring?

Reply:
Yes, for ray.put objects or tasks where max_retries=0 (usually actor tasks).